io-engine/src/bdev/crypto.rs

```diff
@@ -0,0 +1,281 @@
+    let cipher = key_params.cipher.to_string().into_cstring().into_raw();
+    let key = key_params.key.as_str().into_cstring().into_raw();
+    let hex_key2_str = key_params.key2.as_deref();
+    let key2 = if let Some(k2) = hex_key2_str {
+        k2.into_cstring().into_raw()
+    } else {
+        std::ptr::null_mut()
+    };
+    let key_name = key_params.key_name.as_str().into_cstring().into_raw();
+
+    unsafe {
+        debug!(
+            "Creating {} key {} | {k2} of size {}",
+            CStr::from_ptr(cipher).to_str().unwrap(),
+            CStr::from_ptr(key).to_str().unwrap(),
+            key_params.key_len,
+            k2 = if hex_key2_str.is_some() {
+                CStr::from_ptr(key2).to_str().unwrap()
+            } else {
+                ""
+            },
+        );
+    }
+
+    let create_key_params = spdk_accel_crypto_key_create_param {
+        cipher,
+        hex_key: key,
+        hex_key2: key2,
+        key_name,
+        tweak_mode: std::ptr::null_mut(),
+    };
+
+    unsafe {
+        let ret = spdk_accel_crypto_key_create(&create_key_params);
+        if ret != 0 {
+            Err(BdevError::CreateBdevFailed {
+                source: Errno::from_raw(ret.abs()),
+                name: format!("Key create failed: {:?}", create_key_params.key_name),
+            })
+        } else {
+            let key_ptr = spdk_accel_crypto_key_get(create_key_params.key_name);
+            trace!("Key '{:?}' created.", key_params.key_name);
+            Ok(key_ptr)
+        }
+    }
+}
+
+/// callback when operation has been performed on crypto vbdev
+extern "C" fn crypto_vbdev_op_cb(arg: *mut std::os::raw::c_void, errno: i32) {
+    let sender = unsafe { Box::from_raw(arg as *mut oneshot::Sender<i32>) };
+    trace!("crypto_vbdev_op_cb: errno {errno}");
+    sender.send(errno).unwrap();
+}
+
+/// Destroy a crypto vbdev.
+pub async fn destroy_crypto_vbdev(vbdev_name: &str) -> Result<(), BdevError> {
+    let (s, r) = pair::<i32>();
+    unsafe {
+        delete_crypto_disk(
+            vbdev_name.as_ptr() as *mut std::os::raw::c_char,
+            Some(crypto_vbdev_op_cb),
+            cb_arg(s),
+        );
+    }
+
+    r.await
+        .expect("callback gone while deleting crypto disk")
+        .to_result(|e| BdevError::DestroyBdevFailed {
+            source: Errno::from_raw(e),
+            name: vbdev_name.to_string(),
+        })?;
+
+    info!("crypto vbdev destroyed successfully");
+    Ok(())
+}
+/// The primary function to create a crypto vbdev in spdk.
+pub fn create_crypto_vbdev_on_base_bdev(
+    crypto_vbdev_name: &str,
+    base_bdev_name: &str,
+    key_params: &EncryptionKey,
+) -> Result<(), BdevError> {
+    // Create the key.
+    let key = create_crypto_key(key_params)?;
+
+    // Setup the crypto opts using the key now.
+    let crypto_opts_ptr = unsafe {
+        create_crypto_opts_by_name(
+            crypto_vbdev_name.into_cstring().into_raw(),
```

Comment:
same comment here, it's leaking memory, probably as_ptr works here as IIRC this helper functions copies the strings